core/node/provider.go

```diff
@@ -149,13 +151,26 @@ func OfflineProviders() fx.Option {
 	return fx.Provide(provider.NewNoopProvider)
 }
 
-func pinnedProviderStrategy(onlyRoots bool) interface{} {
+func newProvidingStrategy(onlyPinned, onlyRoots bool) interface{} {
 	type input struct {
 		fx.In
 		Pinner      pin.Pinner
+		Blockstore  blockstore.Blockstore
 		IPLDFetcher fetcher.Factory `name:"ipldFetcher"`
 	}
 	return func(in input) provider.KeyChanFunc {
-		return provider.NewPinnedProvider(onlyRoots, in.Pinner, in.IPLDFetcher)
+		if onlyRoots {
+			return provider.NewPinnedProvider(true, in.Pinner, in.IPLDFetcher)
+		}
+
+		var later provider.KeyChanFunc
+		if onlyPinned {
+			later = provider.NewPinnedProvider(false, in.Pinner, in.IPLDFetcher)
```

Comment:
This might be ok, but it might also be pretty unfortunate if the user has tons of tiny pinned objects since it will result in enumerating the datastore twice. Given we're already storing a map with all the pin roots in `NewPrioritizedProvider` maybe we could make something a little more custom for this case like:
- return all recursively pinned roots (and add them to a map)
- return all direct items (and add them to a different map if they're not in the recursive one)
- go through the map of recursively pinned roots and recursively return them all (checking against both maps, and then adding them to the map with the direct items)

Hard for me to say how valuable doing this is, it's a function of how many CIDs are referenced by each pin so users may experience this differently.

Reply:
Addressed in  https://github.com/ipfs/boxo/pull/595/commits/558d9c78084ef50d686c1d415f74a26eea0b8e0c